### PR TITLE
Add `alert-badge` addon

### LINKED ISSUE #6416: Visual indicator on the message badge if there are alerts
### Why this would be helpful

Make it easier to see if you have alerts

### How this addon works

This addon turns the message badge red and adds an exclamation mark when you have alerts

### Possible alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Secret-chest**
Resolves #6416

### Changes

Turns the message badge on the website red if the user has alerts, and adds an exclamation sign. Also adds options to change its colours.

scratch-www
scratchr2

### Reason for changes

To make it easier to see when you have alerts

### Tests

* Chromium 114, Linux Mint 21.2

Also, this is my first true addon (not port)

### TODO

[ ] dynamic enable/disable

### CODE REVIEW

**Comment by Secret-chest** [comment]
@WorldLanguages Is it worth enabling by default?

**Comment by CST1229** [comment]
> Is it worth enabling by default?

Most likely not.

**Comment by Secret-chest** [comment]
> > Is it worth enabling by default?
> 
> Most likely not. (Currently, most addons are disabled by default in SA.)

I was thinking that it's a pretty basic feature, doesn't harm anything and adds no visual clutter.

**Comment by Secret-chest** [comment]
Colours (both default and with alerts) are now also customisable using this addon. Because of this, there is now an option to disable highlighting alerts.

**Comment by Samq64** [comment]
Also, when there's an alert, should it show the full message count or just the number of alerts?

**Comment by WorldLanguages** [comment]
> Also customizable message badge colours were recently added to the website dark mode.

Yeah, that can be a bit problematic. Specially if you consider some people never clear their alerts.

**Comment by WorldLanguages** [comment]
> Is it worth enabling by default?

We *probably* don't want to enable anything by default that has to do with Scratch Team communications.

File: addons/alert-badge/scratchr2.js
[the remark this answers is not in the document]
Comment (Secret-chest) [reply addons/alert-badge/scratchr2.js:1]: What???

**Comment by Secret-chest** [comment]
> Could this be made compatible with editor messages? Also customizable message badge colours were recently added to the website dark mode.

Can you mention the commit? I believe it would be better to have them here.

**Comment by Secret-chest** [comment]
> > Also customizable message badge colours were recently added to the website dark mode.
> 
> Yeah, that can be a bit problematic. Specially if you consider some people never clear their alerts.

It's an addon. I, for example, clear my alerts.

**Comment by Secret-chest** [comment]
> Also, when there's an alert, should it show the full message count or just the number of alerts?

It currently shows the full message count. It is unlikely that you would get very many alerts.

**Comment by Secret-chest** [comment]
> Could this be made compatible with editor messages?

How should I do that?

**Comment by Samq64** [comment]
> > Could this be made compatible with editor messages?
> 
> How should I do that?

The same way as the other pages except it's `.sa-editormessages-count`.

File: addons/alert-badge/scratchr2.js
[the remark this answers is not in the document]
Comment (DNin01) [reply addons/alert-badge/scratchr2.js:1]: Here's an example:

https://github.com/ScratchAddons/ScratchAddons/blob/e997a0edc1eea309f3af6b1ad26feaa41bfe70dd/addons/expanding-search-bar/expanding-search-bar.js#L21

**Comment by Secret-chest** [comment]
> Could this be made compatible with editor messages?
> 
> How should I do that?
> 
> The same way as the other pages except it's .sa-editormessages-count.

OK, I'm doing it tomorrow. It's 22:30 here in Romania.

**Comment by WorldLanguages** [comment]
> It's an addon. I, for example, clear my alerts.

Sure, but you were proposing to enable this by default. If this is disabled by default, then it's okay.

**Comment by Secret-chest** [comment]
> > It's an addon. I, for example, clear my alerts.
> 
> Sure, but you were proposing to enable this by default. If this is disabled by default, then it's okay.

It would be then.

**Comment by Secret-chest** [comment]
> Also customizable message badge colours were recently added to the website dark mode.

Could that be undone? I believe there are too many settings in that addon.

**Comment by WorldLanguages** [comment]
> > Also customizable message badge colours were recently added to the website dark mode.
> 
> Could that be undone? I believe there are too many settings in that addon.

Probably not, as the old orange color is part of the `Scratch default colors (blue)` preset.

**Comment by Secret-chest** [comment]
> Also customizable message badge colours were recently added to the website dark mode.
> 
> Could that be undone? I believe there are too many settings in that addon.
> 
> Probably not, as the old orange color is part of the Scratch default colors (blue) preset.

But you can just use this if you care about such a minor change...

**Comment by Secret-chest** [comment]
> > Also customizable message badge colours were recently added to the website dark mode.
> > Could that be undone? I believe there are too many settings in that addon.
> > Probably not, as the old orange color is part of the Scratch default colors (blue) preset.
> 
> But you can just use this if you care about such a minor change...

Only the one on the messages page was changed. We have better settings to add to website dark mode.

**Comment by WorldLanguages** [comment]
I think before the Scratch colors update it was already planned to add a setting that changed the message count color in the navbar

**Comment by DNin01** [comment]
> > Also customizable message badge colours were recently added to the website dark mode.
> 
> 
> 
> Could that be undone? I believe there are too many settings in that addon.

I get it, but I could imagine it being a highly requested feature.

> But you can just use this if you care about such a minor change...

Sure, but addons dedicated to customizing colors are probably a better place to put such a setting.

Maybe #6425 would help?

**Comment by Secret-chest** [comment]
> > > Also customizable message badge colours were recently added to the website dark mode.
> > 
> > 
> > Could that be undone? I believe there are too many settings in that addon.
> 
> I get it, but I could imagine it being a highly requested feature.
> 
> > But you can just use this if you care about such a minor change...
> 
> Sure, but addons dedicated to customizing colors are probably a better place to put such a setting.
> 
> Maybe #6425 would help?

Well there are already a lot more customisable colours addons, why don't we put block colours in the editor dark mode>

**Comment by Samq64** [comment]
> Well there are already a lot more customisable colours addons, why don't we put block colours in the editor dark mode>

Is your point that message badge colours settings should be part of this addon instead of website dark mode or you want to merge custom block colours and editor dark mode?

**Comment by Secret-chest** [comment]
Addons should be split, so these colours should stay here.

**Comment by DNin01** [comment]
> Addons should be split, so these colours should stay here.

I understand that it would be a more specific place for the messages badge color options to be, but I still don't see why you should have to enable this addon, which has other features, to do so. IMO, putting them in `dark-www` would make them easier to access and even find.

**Comment by Secret-chest** [comment]
Exactly, it makes no sense to have a whole setting inside the dark mode addon for a few pixels on a single page. More important things could go there. Also, the function to highlight alerts can be disabled.

**Comment by DNin01** [comment]
> Exactly, it makes no sense to have a whole setting inside the dark mode addon for a few pixels on a single page. More important things could go there. Also, the function to highlight alerts can be disabled.

We really should've done #6425...

File: addons/alert-badge/scratchr2.js
[the remark this answers is not in the document]
Comment (DNin01) [reply addons/alert-badge/scratchr2.js:1]: I've merged the userscripts into one and fixed a bug or two for you. Mind testing it out?

**Comment by DNin01** [comment]
It's worth considering what should happen when `editor-messages` is disabled. That would mean there is no `.sa-editor-messages` element to find. I haven't fixed that yet. In fact, I might've made it worse (it could create a bunch of pending Promises if the user switches between the project page and the editor enough times due to `await addon.tab.waitForElement(...)`), but that might not be a big deal.

**Comment by Secret-chest** [comment]
Thank you @DNin01

**Comment by DNin01** [comment]
Can someone look into the issue where the messages badge sometimes doesn't have an exclamation point appended or turn red, but then if you reload it sometimes works the second time?

Also, it might be a good idea for consistency with Scratch to make the threshold for black text higher so the original color still has white text. That's how #5602 works.
https://github.com/ScratchAddons/ScratchAddons/blob/ac17dd57440fd7f5317cda67a3a39e83942dc90f/addons/dark-www/addon.json#L1859

**Comment by Secret-chest** [comment]
I though it already did? Please commit the threshold

**Comment by Secret-chest** [comment]
TODO:
[ ] offer options to hide the badge or make it a simple dot
[ ] discuss `editor-dark-mode`

**Comment by DNin01** [comment]
@Secret-chest Pro tip: use `- [ ]` to create a clickable checkbox.

**Comment by Secret-chest** [comment]
> Pro tip: use - [ ] to create a clickable checkbox.

I knew I could, but I forgot you need the dash. I figured it out. Thanks.

**Comment by Secret-chest** [comment]
Should I remove default colour settings from there? Maybe we need addon linking.

**Comment by Secret-chest** [comment]
I removed the default colour changing.

**Comment by Secret-chest** [comment]
Can we merge now? @DNin01 @WorldLanguages @Samq64

File: addons/alert-badge/addon.json [hunk 1]
```diff
@@ -0,0 +1,62 @@
+{
+  "name": "Highlight message badge when there are alerts",
+  "description": "When there are Scratch Team messages, the navbar's notification badge's color is changed and a '!' is added.",
+  "userscripts": [
+    {
+      "url": "userscript.js",
+      "matches": ["projects", "scratchWWWNoProject", "isNotScratchWWW"]
```
Comment (Samq64) [inline addons/alert-badge/addon.json:7]:
```suggestion
      "matches": ["https://scratch.mit.edu/*"]
```

File: addons/alert-badge/addon.json [hunk 2]
```diff
@@ -0,0 +1,62 @@
+{
+  "name": "Highlight message badge when there are alerts",
+  "description": "When there are Scratch Team messages, the navbar's notification badge's color is changed and a '!' is added.",
```
Comment (Samq64) [inline addons/alert-badge/addon.json:3]:
```suggestion
  "description": "When there are Scratch Team messages, the color of the notification badge in the navigation bar is changed and an exclamation mark is added.",
```

File: addons/alert-badge/addon.json [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+      "matches": ["projects", "scratchWWWNoProject"]
+    },
+    {
+      "url": "scratchr2.css",
+      "matches": ["isNotScratchWWW"]
+    },
+    {
+      "url": "editor.css",
+      "matches": ["projects"]
+    }
```
Comment (Samq64) [inline addons/alert-badge/addon.json:22]: The userscript handles adding the class so could the userstyles be merged?

File: addons/alert-badge/addon.json [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+  ],
+  "injectAsStyleElt": true,
+  "settings": [
+    {
+      "name": "Badge color when there are alerts",
```
Comment (Samq64) [inline addons/alert-badge/addon.json:27]:
```suggestion
      "name": "Alert badge color",
```

File: addons/alert-badge/addon.json [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+      "id": "alert",
+      "type": "color",
+      "default": "#cc4400",
+      "if": {
+        "settings": { "highlight": true }
```
Comment (Samq64) [inline addons/alert-badge/addon.json:32]: This setting doesn't exist any more.

File: addons/alert-badge/addon.json [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+      }
+    }
+  ],
+  "customCssVariables": [
+    {
+      "name": "alert",
+      "value": { "type": "settingValue", "settingId": "alert" }
```
Comment (Samq64) [inline addons/alert-badge/addon.json:39]: Isn't this just the same as the setting's value?

**Review by Samq64 — COMMENTED** [review]
Have you checked if the `!important`s are actually still necessary?

**Comment by Waakul** [comment]
> Have you checked if the `!important`s are actually still necessary?

Things that need !important in other userstyle managers dont need them in scratchaddons, sa probably automates the prioritizing process with other methods.

**Comment by Secret-chest** [comment]
> Have you checked if the !importants are actually still necessary?

Yes. I have tried it without.

**Comment by Secret-chest** [comment]
@Samq64 can you please do the changes you suggested? I haven't got a testing environment

**Comment by Samq64** [comment]
I'll need to look into this a little more. The message badge disappears when going back to the project page and I tried to add `runAtComplete: false` but it seemed to cause a race condition on `scratch-www` even with an `addon.tab.waitForElement`.

**Comment by Samq64** [comment]
> The message badge disappears when going back to the project

It seems to be a Scratch bug.

`querySelector` and `waitForElement` support multiple selectors so the while true waitForElement method could simply things here.

**Pull request closed without merging**